crates/storage/src/index_map.rs

```diff
@@ -0,0 +1,430 @@
+    where
+        T: 'b,
+    {
+        self.no_prefix().keys_raw(store, min, max, order)
+    }
+
+    pub fn keys<'b>(
+        &self,
+        storage: &'b dyn Storage,
+        min: Option<Bound<K>>,
+        max: Option<Bound<K>>,
+        order: Order,
+    ) -> Box<dyn Iterator<Item = StdResult<K::Output>> + 'b> {
+        self.no_prefix().keys(storage, min, max, order)
+    }
+
+    pub fn clear(
+        &self,
+        storage: &mut dyn Storage,
+        min: Option<Bound<K>>,
+        max: Option<Bound<K>>,
+        limit: Option<usize>,
+    ) {
+        self.no_prefix().clear(storage, min, max, limit)
+    }
+
+    pub fn range<'b>(
+        &self,
+        storage: &'b dyn Storage,
+        min: Option<Bound<K>>,
+        max: Option<Bound<K>>,
+        order: Order,
+    ) -> Box<dyn Iterator<Item = StdResult<(K::Output, T)>> + 'b> {
+        self.no_prefix().range(storage, min, max, order)
+    }
+}
+
+impl<'a, K, T, I, E: Encoding<T>> IndexedMap<'a, K, T, I, E>
+where
+    K: MapKey,
+    E: Encoding<T>,
+{
+    pub fn load(&self, storage: &dyn Storage, key: K) -> StdResult<T> {
+        self.primary.load(storage, key)
+    }
+
+    pub fn may_load(&self, storage: &dyn Storage, key: K) -> StdResult<Option<T>> {
+        self.primary.may_load(storage, key)
+    }
+}
+
+impl<'a, K, T, I, E: Encoding<T>> IndexedMap<'a, K, T, I, E>
+where
+    K: MapKey + Clone,
+    I: IndexList<T>,
+    E: Encoding<T>,
+{
+    pub fn save(&'a self, storage: &mut dyn Storage, key: K, data: &T) -> StdResult<()> {
+        let old_data = self.may_load(storage, key.clone())?;
+        self.replace(storage, key, Some(data), old_data.as_ref())
+    }
+
+    pub fn remove(&'a self, storage: &mut dyn Storage, key: K) -> StdResult<()> {
+        let old_data = self.may_load(storage, key.clone())?;
+        self.replace(storage, key, None, old_data.as_ref())
+    }
+
+    pub fn replace(
+        &'a self,
+        storage: &mut dyn Storage,
+        key: K,
+        data: Option<&T>,
+        old_data: Option<&T>,
+    ) -> StdResult<()> {
+        // this is the key *relative* to the primary map namespace
+        let pk = key.serialize();
+        if let Some(old) = old_data {
+            for index in self.idx.get_indexes() {
+                index.remove(storage, &pk, old)?;
+            }
+        }
+        if let Some(updated) = data {
+            for index in self.idx.get_indexes() {
+                index.save(storage, &pk, updated)?;
+            }
+            self.primary.save(storage, key, updated)?;
+        } else {
+            self.primary.remove(storage, key);
+        }
+        Ok(())
+    }
+}
+
+impl<'a, K, T, I, E: Encoding<T>> IndexedMap<'a, K, T, I, E>
+where
+    K: MapKey + Clone,
+    I: IndexList<T>,
+    T: Clone,
+    E: Encoding<T>,
+{
+    pub fn update<A, Err>(&'a self, storage: &mut dyn Storage, key: K, action: A) -> Result<T, Err>
+    where
+        A: FnOnce(Option<T>) -> Result<T, Err>,
```

Comment:
We changed the `update` function signature from CosmWasm's. Now the `action` function should a `Result<Option<T>, Err>`. Returning a `None` means the data is to be deleted.